### ISSUE TITLE
Deactivate idle inhibitor programmatically

### ISSUE THREAD

**Issue by adriangonz** [Alexays/Waybar#690]
It would be good to have a way to active / deactivate the idle inhibitor module programmatically (e.g. from a CLI).

The use I have in mind is that whenever I lock my laptop, if I forget to deactivate the idle inhibitor on Waybar, it will just never go to sleep. Thus, it would be great for me if I could just deactivate it every time I lock my laptop.

I checked the source code, and it seems that this is currently implemented using something from `wlroots`? There may be some way to interact with that already, but I haven't found any.

**Comment by kmARC** [comment]
I'd love to see this nevertheless; My use cases would be:

* upon manual locking, I would like to disable the inhibitor for the future
* After 2 hours (or whatever longer period) I would like to get my screen locked anyways

Both these are needed since I "forget" that the inhibitor is on

**Comment by arnej** [comment]
I would also be interested in this.

My use case is:
When I remove my YubiKey, I run several things. One of them is to run swaylock to lock the screen. When idle inhibitor is still active, the lock works fine, but the screen is never turned off. I would like to first disable the idle inhibitor before I lock the screen so that the screen could be turned off as if it was locked automatically.

**Comment by lunik1** [comment]
I would like something like this too, along with a way to trigger the idle inhibitor manually with a keyboard shortcut.

I thought one possibility for this would be to wait for a signal, but `SIGUSR1` and `SIGUSR2` are already in use...

**Comment by 3nuc** [comment]
This would be useful for eg. watching video on a computer with KDE Connect.
1. Play video
2. Lay down on other side of room
3. Ah I forgot to turn on idle inhibitor
4. Use KDE Connect "Run Command" to enable idle inhibitor

**Comment by JohnMertz** [comment]
I'd also be very interested in this. My main concern is to be able to restore the idle state from the last time Waybar was running. I have a script to automatically reconfigure displays when I dock/undock and part of that involves killing and then load a different Waybar config. This always deactivates the idle_inhibitor and my memory is such that at least 80% of the time I end up getting swayidle notifications about impending locking before I re-enable it.

I'd just implement a fix for this by having an 'on-click' event which calls a script that writes the new state to text file, then after whatever process starts Waybar, I'd just have the state in that text file restored.

This is literally the only complaint I've had with the Waybar in a couple years of using it.

**Comment by nimitzpro** [comment]
@JohnMertz I'm currently starting and killing the swayidle process by running a script with on-click (which could extend to read/write to a file), but how do you restore state from the file back to waybar?

**Comment by JohnMertz** [comment]
@nimitzpro I hadn't had a way to restore the state from the file when waybar starts at the time I last replied. I was just joining the plea to have this issue addressed. However, since then I've basically figured out a workaround...

Technically I start `swayidle` and then never actually use the `idle_inhibitor`, I just change the idle behaviour based on the state stored in the file. It always starts with a list of actions, but depending on the state logged in the file, it will just ignore certain actions when the specific timeout is reached, up to and including not doing anything at all (effectively applying the `idle_inhiditor` by bypassing all idle actions).

Here's [the shell script which starts `swayidle`, and executes each individual action](https://git.john.me.tz/jpm/scripts/src/branch/master/sway/idle.sh) and [the script that is run by waybar to get the current state and rotate through the available idle states](https://git.john.me.tz/jpm/scripts/src/branch/master/waybar/waybar-idle.sh)

Note: the `sleep` state doesn't work properly. It does not execute until `swaylock` is authenticated, then it shuts off the output and immediately comes back. 99% of the time I just use `lock`, so I've been too lazy to sort this out. The `fade` mode is also fairly boutique for my setup,

**Comment by coreyoconnor** [comment]
I wonder if this is the right place for programmatic idle inhibit. Should waybar, instead, reflect the current idle inhibitor lock state and rely on other wayland protocol clients to do the complex use cases? 

The underlying wayland protocol: https://wayland.app/protocols/idle-inhibit-unstable-v1#zwp_idle_inhibit_manager_v1

does not have a way for an application to monitor the idle inhibitors. As far as I can tell anyways; I'm definitely not knowledable about wayland!

If that protocol did have a way then some of the programmatic change of idle inhibit could be separate applications. Waybar module would then listen for idle inhibit changes and reflect that state in the internal module.

**Comment by machitgarha** [comment]
Doesn't Waybar has an IPC option? So why not using it for this?

**Comment by nimitzpro** [comment]
> I wonder if this is the right place for programmatic idle inhibit. Should waybar, instead, reflect the current idle inhibitor lock state and rely on other wayland protocol clients to do the complex use cases? 

I have been able to alter the idle inhibit using a custom tab in waybar that changes a text file or runs a script, (used to use swayidle, now on hypridle).

I need to look into whether there's a way to show the current idle state if it has been altered by something else (e.g. a script, config, or macro). For example Hyprland auto inhibits sleeps when some windows are focused or active.

**Comment by CD11b** [comment]
Using `systemctl restart --user waybar.service` seems to be a workaround for now.

**Comment by guttermonk** [comment]
Created a PR to close this out and add another feature: https://github.com/Alexays/Waybar/pull/4565

**Comment by iyzana** [comment]
@guttermonk A signal to _toggle_ the idle inhibitor will not solve the original use case of this issue. With a toggle it's still not possible to reliably _deactivate_ it before locking

**Comment by guttermonk** [comment]
@iyzana, you're right. It works with a keybind, but it surprisingly doesn't seem to work when binding it to hyprland's lid switch like you said. That's too bad.

**Comment by iyzana** [comment]
That's not really what I meant. If there's only a signal to _toggle_ the idle inhibitor, then there's no way to prevent it from turning _ON_ when locking if it was already off, right?

**Comment by guttermonk** [comment]
You can try using a script with a simple "if activated, then `pkill -SIGRTMIN+6 waybar`" that would execute with the lid switch.

I use swayidle, and this worked for me:
```
#!/usr/bin/env bash

# Only toggle waybar's idle_inhibitor if it's currently activated
# Detects state by checking if swayidle is being inhibited

check_idle_inhibited() {
    # Check if swayidle is running
    if ! pgrep swayidle > /dev/null; then
        echo "Warning: swayidle not running, cannot detect inhibitor state"
        return 2
    fi
    
    # Test if idle is being inhibited
    TEST_FILE="/tmp/waybar_caffeine_test_$$"
    
    # Start a test swayidle instance with 1 second timeout
    timeout 2.5 swayidle -d \
        timeout 1 "touch $TEST_FILE" \
        resume "rm -f $TEST_FILE" >/dev/null 2>&1 &
    
    TEST_PID=$!
    
    # Wait for potential idle trigger
    sleep 2
    
    # Clean up test process
    kill $TEST_PID 2>/dev/null
    
    # Check if idle was triggered
    if [ -f "$TEST_FILE" ]; then
        # Idle triggered = NOT inhibited
        rm -f "$TEST_FILE"
        return 1
    else
        # Idle prevented = IS inhibited
        return 0
    fi
}

# Check if idle_inhibitor is currently active
if check_idle_inhibited; then
    echo "Caffeine is active - toggling off"
    pkill -SIGRTMIN+6 waybar
    notify-send "Caffeine" "Deactivated ☕" -t 2000
else
    echo "Caffeine is already inactive - doing nothing"
    notify-send "Caffeine" "Already off" -t 2000
fi
```

Note: it may not have worked for me earlier since I had `start-activated = true;` and I haven't really had this issue so it's hard to test.

**Comment by tremby** [comment]
My use case is a script that turns on "presentation mode" and locks it that way until killed. It does various things like changing performance mode, setting a known constant audio volume level, turning off notifications, and I want it to turn on the inhibitor too.

I could probably do this with `systemd-inhibit` but it'd be nice to do it via waybar so that it's reflected in the UI (giving confidence that it's working), and it'd also allow me to do one-shot "turn on presentation mode" and "turn off presentation mode" commands rather than only having a "on until script killed" mode.

**Comment by iyzana** [comment]
@Alexays there is still no way to reliably disable the idle inhibitor if there is only a signal to toggle it

**Comment by tremby** [comment]
@iyzana, maybe it's possible if there's a way to query whether it's currently on or not. But I'm not sure if there is such a thing.

If not @Alexays, please add a way to force it off and force it on, instead of or in addition to the toggle.
